msprime/trees.py

```diff
@@ -371,6 +381,22 @@ def _assign_x_coordinates(self, node):
             self._x_coords[node] = self._leaf_x * self._x_scale
             self._leaf_x += 1
 
+    def _assign_colours(self, node):
+        """
+        Assign colours up the tree, making parents the same
+        colour as their children, unless there is a disagreement
+        in which case no colour is set
+        """
+        if self._tree.is_internal(node):
+            children = self._tree.get_children(node)
+            for c in children:
+                self._assign_colours(c)
+            if node not in self._colour:
+                colours = set([self._colour.get(c) for c in children])
```

Comment:
The assumption here is that not all nodes need to have a colour. So self._colour[c] may fail, in which case we simply treat it as colour=None, and it does not get a colour assigned. In the sag, that means it simply inherits the colour from the line group, rather than having an explicit override.